browser/ui/tabs/shared_pinned_tab_service.cc

```diff
@@ -0,0 +1,680 @@
+      continue;
+    }
+
+    tab_strip_model->UpdateWebContentsStateAt(index, change_type);
+  }
+}
+
+void SharedPinnedTabService::OnProfileWillBeDestroyed(Profile* profile) {
+  profile_will_be_destroyed_ = true;
+}
+
+void SharedPinnedTabService::OnTabAdded(
+    TabStripModel* tab_strip_model,
+    const TabStripModelChange::Insert* insert) {
+  DVLOG(2) << __FUNCTION__;
+  LOCK_REENTRANCE(tab_strip_model);
+  DCHECK(insert);
+
+  for (const auto& contents_with_index : insert->contents) {
+    auto current_index =
+        tab_strip_model->GetIndexOfWebContents(contents_with_index.contents);
+    if (!tab_strip_model->IsTabPinned(current_index)) {
+      continue;
+    }
+
+    auto tab_renderer_data =
+        TabRendererData::FromTabInModel(tab_strip_model, current_index);
+    if (static_cast<int>(pinned_tab_data_.size()) > current_index &&
+        pinned_tab_data_.at(current_index).renderer_data.last_committed_url ==
+            tab_renderer_data.last_committed_url) {
+      DummyContentsData::CreateForWebContents(
+          /* dummy_contents= */ contents_with_index.contents,
+          /* shared_contents= */ pinned_tab_data_.at(current_index)
+              .shared_contents);
+      continue;
+    }
+
+    SharedContentsData::CreateForWebContents(contents_with_index.contents);
+    DCHECK_LE(contents_with_index.index,
+              static_cast<int>(pinned_tab_data_.size()));
+    pinned_tab_data_.insert(
+        pinned_tab_data_.begin() + contents_with_index.index,
+        {.renderer_data = tab_renderer_data,
+         .shared_contents = contents_with_index.contents,
+         .contents_owner_model = tab_strip_model});
+
+    SynchronizeNewPinnedTab(contents_with_index.index);
+  }
+}
+
+void SharedPinnedTabService::OnTabMoved(TabStripModel* tab_strip_model,
+                                        const TabStripModelChange::Move* move) {
+  if (!tab_strip_model->IsTabPinned(move->to_index)) {
+    return;
+  }
+
+  if (static_cast<int>(pinned_tab_data_.size()) <= move->from_index) {
+    // This case is where a tab is newly pinned. This case will be dealt in
+    // TabPinnedStateChanged().
+    return;
+  }
+
+  DVLOG(2) << __FUNCTION__;
+
+  LOCK_REENTRANCE(tab_strip_model);
+
+  // Now we're handling a pinned tab's was moved while remaining pinned.
+  if (move->from_index < move->to_index) {
+    std::rotate(pinned_tab_data_.begin() + move->from_index,
+                pinned_tab_data_.begin() + move->from_index + 1,
+                pinned_tab_data_.begin() + move->to_index + 1);
+  } else {
+    std::rotate(pinned_tab_data_.begin() + move->to_index,
+                pinned_tab_data_.begin() + move->from_index,
+                pinned_tab_data_.begin() + move->from_index + 1);
+  }
+
+  SynchronizeMovedPinnedTab(move->from_index, move->to_index);
+}
+
+void SharedPinnedTabService::OnTabRemoved(
+    TabStripModel* tab_strip_model,
+    const TabStripModelChange::Remove* remove) {
+  if (last_active_browser_ &&
```

Comment:
This is because when we unpin a tab on the active browser, it removes the tab from all the other browsers, right?